File: lib/JFFUtils/NSString/NSString+PathExtensions.m
#import "NSString+PathExtensions.h"

@implementation NSString (PathExtensions)

+(NSString*)pathForDirectory:(NSSearchPathDirectory)directory
{
    NSArray  *pathes = NSSearchPathForDirectoriesInDomains(directory, NSUserDomainMask, YES);
    return [pathes lastObject];
}

+ (NSString *)documentsPathByAppendingPathComponent:(NSString *)str
{
    static NSString *documentsDirectory = nil;
    
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken,
    ^{
        documentsDirectory = [ self pathForDirectory: NSDocumentDirectory ];
    });
    
    return [ documentsDirectory stringByAppendingPathComponent: str ];
}

+ (NSString *)cachesPathByAppendingPathComponent:(NSString *)str
{
    static NSString *cachesDirectory = nil;
    
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken,
    ^{
       cachesDirectory = [ self pathForDirectory: NSCachesDirectory ];
    });
    
    return [ cachesDirectory stringByAppendingPathComponent: str ];
}

@end

File: iAsync.Swift/lib/JUtils/JUtils/String/NSString+PathExtensions.swift
//
//  NSString+PathExtensions.swift
//  JUtils
//
//  Created by Vlafimir Gorbenko on 06.06.14.
//  Copyright (c) 2014 EmbeddedSources. All rights reserved.
//

import Foundation

public extension NSString {
    
    private class func pathWithSearchDirecory(directory: NSSearchPathDirectory) -> String {
        
        let pathes = NSSearchPathForDirectoriesInDomains(directory, .UserDomainMask, true) as! [String]
        return pathes[pathes.endIndex - 1]
    }
    
    class func documentsPathByAppendingPathComponent(str: String) -> String {
        
        struct Static {
            static var instance = NSString.pathWithSearchDirecory(.DocumentDirectory)
        }
        
        return Static.instance.stringByAppendingPathComponent(str)
    }
    
    class func cachesPathByAppendingPathComponent(str: String) -> String {
        
        struct Static {
            static var instance = NSString.pathWithSearchDirecory(.CachesDirectory)
        }
        
        return Static.instance.stringByAppendingPathComponent(str)
    }
}
